pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -90,22 +125,335 @@ func New(cfg Config,
+		// so more than one index storage client may end up listing the same set of buckets
+		// avoid including the same table twice in the compact tables list.
+		seen = make(map[string]struct{})
+	)
+	for _, sc := range c.storeClients {
+		// refresh index list cache since previous compaction would have changed the index files in the object store
+		sc.index.RefreshIndexTableNamesCache(ctx)
+		tbls, err := sc.index.ListTables(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to list tables: %w", err)
+		}
+
+		for _, table := range tbls {
+			if _, ok := seen[table]; ok {
+				continue
+			}
+
+			tables = append(tables, table)
+			seen[table] = struct{}{}
+		}
+	}
+
+	// process most recent tables first
+	sortTablesByRange(tables)
+
+	// apply passed in compaction limits
+	if c.cfg.SkipLatestNTables <= len(tables) {
+		tables = tables[c.cfg.SkipLatestNTables:]
+	}
+	if c.cfg.TablesToCompact > 0 && c.cfg.TablesToCompact < len(tables) {
+		tables = tables[:c.cfg.TablesToCompact]
+	}
+
+	// Reset discovered tenants metric since we will increase it in compactTable.
+	c.metrics.compactionRunDiscoveredTenants.Set(0)
+
+	parallelism := c.cfg.MaxCompactionParallelism
+	if parallelism == 0 {
+		parallelism = len(tables)
+	}
+
+	errs := multierror.New()
+	if err := concurrency.ForEachJob(ctx, len(tables), parallelism, func(ctx context.Context, i int) error {
+		tableName := tables[i]
+		level.Info(util_log.Logger).Log("msg", "compacting table", "table-name", tableName)
+		err := c.compactTable(ctx, tableName)
+		if err != nil {
+			errs.Add(err)
+			return nil
+		}
+		level.Info(util_log.Logger).Log("msg", "finished compacting table", "table-name", tableName)
+		return nil
+	}); err != nil {
+		errs.Add(err)
+	}
+
+	return errs.Err()
+}
+
+func (c *Compactor) compactTable(ctx context.Context, tableName string) error {
+	schemaCfg, ok := schemaPeriodForTable(c.schemaCfg, tableName)
+	if !ok {
+		level.Error(util_log.Logger).Log("msg", "skipping compaction since we can't find schema for table", "table", tableName)
+		return nil
+	}
+
+	sc, ok := c.storeClients[schemaCfg.From]
+	if !ok {
+		return fmt.Errorf("index store client not found for period starting at %s", schemaCfg.From.String())
+	}
+
+	_, tenants, err := sc.index.ListFiles(ctx, tableName, false)
+	if err != nil {
+		return fmt.Errorf("failed to list files for table %s: %w", tableName, err)
+	}
+
+	c.metrics.compactionRunDiscoveredTenants.Add(float64(len(tenants)))
+	level.Info(c.logger).Log("msg", "discovered tenants from bucket", "users", len(tenants))
+	return c.compactUsers(ctx, sc, tableName, tenants)
+}
+
+// See: https://github.com/grafana/mimir/blob/34852137c332d4050e53128481f4f6417daee91e/pkg/compactor/compactor.go#L566-L689
+func (c *Compactor) compactUsers(ctx context.Context, sc storeClient, tableName string, tenants []string) error {
+	// When starting multiple compactor replicas nearly at the same time, running in a cluster with
```

Comment:
I don't think this really applies here. The tsdb compactor is currently run as a singleton, meaning it compacts everything from one process. This code was a safeguard to minimize duplicate work when multiple compactors were erroneously configured.

In the bloom compactor, we expect to run multiple compactors in a ring. Each replica should operate on different data, making this safeguard unnecessary.

There is one exception: when the ring topology changes and the compactor replicas have different views of ring membership when determining what data to compact.

In summary, this shouldn't help us much, but also doesn't really hurt. Just wanted to make sure you understood why this optimization isn't necessary for the bloom compactors compared to the tsdb compactor.

Reply:
That makes sense, thank you! I removed that shuffling.